### DISCUSSION TITLE
A Minimal Rust Kernel (ru)

### DISCUSSION THREAD

**Discussion by giscus** [phil-opp/blog_os#1044] [Post Comments (translated)]
# A Minimal Rust Kernel (ru)

В этом посте мы создадим минимальное 64-битное ядро на Rust для архитектуры x86_64. Мы будем отталкиваться от независимого бинарного файла из предыдущ…

https://os.phil-opp.com/ru/minimal-rust-kernel/

**Comment by Exiled1** [comment]
Hey! Incredible tutorial! Also, for anyone else following along, I tried to fix a weird bootimage error with the bootloader 0.10.X crate, and it turns out that the bootimage crate isn't used in 0.10.X, so use the bootloader 0.9.8 version instead until they talk about upgrading :)

**Comment by happlex** [comment]
I just rewrote the `_start()` program in functional way!  : )

```rust
#[no_mangle]
pub extern "C" fn _start() -> ! {
    let vga_buffer = 0xb8000 as *mut u8;

    "Hello, world!"
        .as_bytes()
        .iter()
        .flat_map(|bt| [*bt, 0xf as u8])
        .enumerate()
        .for_each(|(i, byte)| unsafe {
            *vga_buffer.offset(i as isize) = byte;
        });

    loop {}
}
```

**Comment by h3nry-d1az** [comment]
Seriously bro, I love your tutorials and I am learning a lot from them, keep it up because you are the best. You deserve to have much more recognition within the tutorials to learn how to develop operating systems. 👍

**Reply by phil-opp** [reply]
Thanks so much!

**Comment by jamincan** [comment]
I can't seem to get the binary to run in QEMU, ending up with just a blank screen. It appear that the BIOS loading bit flashes up briefly, but too quickly to see. I thought maybe a typo could be causing the problem, but even checking out post-02 from the blog_os repo results in the same outcome. Do you have any ideas of where to start troubleshooting?

**Reply by bjorn3** [reply]
Are you using QEMU 6.0.0? If so try updating to 6.1.0 (or later): https://github.com/phil-opp/blog_os/issues/1052

**Reply by goodman965** [reply]
I have the same problem, qemu update didn't help me. I tried to run the Linux version of the emulator in a virtual machine, but the same is there black screen. I checked all code, preferences, versions, everything the same as here but not working.

**Reply by goodman965** [reply]
Finally I found a mistake, a trivial typo in this line:
`const BUFFER_HEIGHT: usize = 26;`

**Comment by jamincan** [comment]
That seems to have been the problem; thanks!

**Comment by Jakobohrnell** [comment]
having issues after setting up the QEMU runner.

This is my config.toml
```
[unstable]
build-std-features = ["compiler-builtins-mem"]
build-std = ["core", "compiler_builtins"]

[target.'cfg(target_os = "none")']
runner = "bootimage runner"
```
and I'm getting this error
```
Error: Failed to execute QEMU run command `"qemu-system-x86_64" "-drive" "format=raw,file=target\\x86_64-prot_os\\debug\\bootimage-proto_os.bin"`: An I/O error occured: The system cannot find the file specified. (os error 2)
```
Seems to be a relative path error?

**Reply by bjorn3** [reply]
Does running `qemu-system-x86_64` work? If not you will need to add the directory that contains the `qemu-system-x86_64` you installed to `PATH`.

**Reply by Jakobohrnell** [reply]
Thank you! Just addig qemu to path worked :)

**Comment by denzyldick** [comment]
Nice it works. Thanks.

**Comment by aniketfuryrocks** [comment]
op. You should sell a hard-copy of this book.

**Comment by JakkuSakura** [comment]
On the latest nightly rustc, bootloader 0.9.19 should be used, as 0.9.18 uses some removed features

**Comment by metsis** [comment]
awesome stuff. thank you, it was all very clear and concise. and worked on the first try! :)

**Reply by phil-opp** [reply]
Great to hear that, thanks!

**Comment by DIngDONG230** [comment]
whenever I run the qemu command "qemu-system-x86_64 -drive format=raw,file=target/x86_64-blog_os/debug/bootimage-blog_os.bin" it gives me in the qemu some weird text

**Reply by bjorn3** [reply]
Can you post a screenshot?

**Comment by sonyynosification** [comment]
All clean and concise, carefully explained and documented. Thank you so much.

**Comment by Senrelio** [comment]
Really need some help to follow this amazing series.

qemu command gives me this error message:
```
Failed to initialize module: /usr/lib/x86_64-linux-gnu/qemu/ui-gtk.so
Note: only modules from the same build can be loaded.
```
I'm sure my PC is x86_64, and it's the same result when run qemu on github's post-03 branch. I assume it has something to do with the x86_64-linux-gnu and x86_64-unknown-none triples.

**Reply by bjorn3** [reply]
That looks like a corrupt qemu installation. Try installing the `qemu-system-gui` package.

**Reply by Senrelio** [reply]
Thanks!

I fixed it by removing already installed `qemu-system-gui` and `qemu-system`, then reinstalled `qemu-system-gui`.

**Comment by tomabailey** [comment]
This is so beautifully written and easy to follow. I’m new to rust, and I’m new to even thinking about OS development so please bear with me if this is a daft question: what sort of hardware would be suitable for this sort of project? I’m guessing something like a RPi would work, but could I go smaller and develop an OS like this on ARM based microcontroller development boards? Any guidance welcome!

**Reply by phil-opp** [reply]
The series targets the x86_64 architecture, i.e. the Intel/AMD CPUs that are found in most PCs.

**Comment by cantrepro** [comment]
I think I followed this guide to the letter but I get this error when doing `cargo build`:
```   
Compiling x86_64 v0.13.2
error: cannot find macro `asm` in this scope
...
error: could not compile `x86_64` due to 42 previous errors
Error: Bootloader build failed.
```
(All 42 errors are the same `cannot find macro asm`)

Any ideas as to what is going wrong here?

**Reply by bjorn3** [reply]
The inline asm stabilization PR (https://github.com/rust-lang/rust/pull/91728) changed the location where the `asm!()` macro can be found. The x86_64 crate hasn't yet been updated for this. There is a pending PR: https://github.com/rust-osdev/x86_64/pull/329 For now try using a rustc nightly from before the stabilization of inline asm like nightly-2021-12-13.

**Reply by cantrepro** [reply]
Thank you!

**Comment by baiyulong** [comment]
[dependencies]
bootloader = "0.10.11"

When I run "cargo build'. I got blew error

Error: An error occured while trying to build the bootloader: The `bootloader` dependency has not the right format: No `package.metadata.bootloader.target` key found in Cargo.toml of bootloader

(If you're using the official bootloader crate, you need at least version 0.5.1)

Caused by:
    The `bootloader` dependency has not the right format: No `package.metadata.bootloader.target` key found in Cargo.toml of bootloader
    
    (If you're using the official bootloader crate, you need at least version 0.5.1)

**Reply by phil-opp** [reply]
The guide still uses version 0.9 of the `bootloader` crate. Version 0.10 includes some breaking changes and is not directly compatible. So just specify version 0.9 in your Cargo.toml for now.

**Comment by wil-den** [comment]
Hi,
Like most others, I think the series is great. I am working my way through it.
I have a question about target triples. 
So if i wanted to target a different system like arm and lets say I have a specific board (in this case the one I have here at home is a TM4C123GXL Tiva C Evaluation kit) what properties of the triple specified here would I need to change to in theory get it to run? I appreciate I probably wouldn't be able to use the VGA Text buffer either. 

Thanks in advance,
Wil

**Reply by phil-opp** [reply]
Looks like that board has an ARM Cortex M4F CPU, so you should be able to use the built-in `thumbv7em-none-eabihf` of `rustc`. So you don't need to create a target JSON file. If you're still interested in the details of the target, you can use the following command to print the corresponding JSON:

```
rustc +nightly --print target-spec-json -Z unstable-options --target thumbv7em-none-eabihf
```

The Cortex-M cores don't have a memory management unit (MMU), so they don't support different virtual address spaces. This means that creating page tables and different processes is not really possible on these CPUs. So instead of a normal operating system, they typically run some more rudimentary OS or even run the programs directly on the metal. This environment is called an "embedded system".

Since code for embedded systems often differs significantly from general purpose operating system code, this tutorial here probably won't be that useful for Cortex-M development. However, there is a great introduction to embedded development in Rust at <https://docs.rust-embedded.org/book/>. It uses the Cortex-M architecture so you should be able to follow along. See also the [`cortex-m-quickstart`](https://github.com/rust-embedded/cortex-m-quickstart) project, which should give you a good starting point.

**Reply by wil-den** [reply]
Thanks for taking the time for the lengthy reply. The rustc command is what I am really looking for to compare the different triples based on system type.
Thanks again.
I was aware of the embedded book, just havent been able to get to it just yet. But I think I'll give it a go.

Wil

**Comment by zendevil** [comment]
Hi
setting target in cargo.toml and running `cargo build` gives me:
error: -Zbuild-std requires --target

**Reply by phil-opp** [reply]
You need to set the default target in a `.cargo/config.toml` file, not in the `Cargo.toml`.

**Comment by LukeMcAuleyDublin** [comment]
Awesome tutorial man, thank you!

**Comment by aashwinr** [comment]
First of all great tutorial!

`Error: Failed to execute QEMU run command `"qemu-system-x86_64" "-drive" "format=raw,file=target/x86_64-vallus_os/debug/bootimage-vallus_os.bin"`: An I/O error occured: No such file or directory (os error 2)`

I'm getting this error when I try to run cargo r. I'm able to run it normally when I type in the full path to the binary (Installed Qemu using brew so it's in my path).

**Reply by aashwinr** [reply]
Sorry for the bad formatting!

**Reply by bjorn3** [reply]
Does `qemu-system-x86_64` pop up a qemu window?

**Comment by mooxiu** [comment]
Thank you for your wonderful post!
However, after setting the default target under `.cargo/config.toml`, when running
```
cargo install bootimage
```
it would try to find the default target `x86_64-blig_os.json` and  return an error:
```
error: target path "/Users/username/x86_64-blog_os.json" is not a valid file
```

**Reply by bjorn3** [reply]
Did you place `x86_64-blog_os.json` in your project next to `Cargo.toml`?

**Comment by EGL20FL** [comment]
Well done!

**Comment by grawitti** [comment]
HI!
I built the 'post-02' branch using:
rustc 1.81.0-nightly (0c81f94b9 2024-07-10)
host: x86_64-unknown-linux-gnu
release: 1.81.0-nightly
LLVM version: 18.1.7
QEMU emulator version 9.0.1

After running qemu it outputs to the console: `VNC server running on ::1:5900`
and nothing else, no `"Hello World!"` line output.

**Reply by grawitti** [reply]
I use the `cargo run` command to run qemu.
Also i try run qemu with command: `qemu-system-x86_64 -drive format=raw,file=target/x86_64-blog_os/debug/bootimage-blog_os.bin`.

**Reply by phil-opp** [reply]
Did you change your QEMU config in any way?

**Reply by grawitti** [reply]
No, I have not changed the qemu configuration.

**Reply by grawitti** [reply]
Solved: Using virt-viewer, I connected to the vnc server vnc://localhost:5900, the window opened and displayed the string "Hello World!". Thanks!
